securesystemslib/metadata.py

```diff
@@ -97,3 +101,90 @@ def pae(self) -> bytes:
             len(self.payload),
             self.payload,
         )
+
+    def sign(self, signer: Signer) -> Signature:
+        """sign the payload and create the signature
+
+        Arguments:
+            signer: A "Signer" class instance
+
+        Raises:
+            TypeError: If "signer" is not a instance of "Signer" class.
+
+        Returns:
+            A "Signature" instance.
+        """
+
+        if not isinstance(signer, Signer):
+            raise TypeError(f"expected type 'Signer', got {type(signer).__name__}")
+
+        signature = signer.sign(self.pae)
+        self.signatures.append(signature)
+
+        return signature
+
+    def verify(self, keys: KeyList, threshold: int) -> List[str]:
+        """verify payload with multiple verifiers.
+
+        Arguments:
+            keys: A list of a "str" and "Key" class instance.
+            threshold: required no of signatures have to pass the verification.
+
+        Raises:
+            TypeError: If key in "keys" is not a instance of "Key" class.
+            ValueError: If "threshold" is not valid.
+            SignatureVerificationError: If "recognized_signers" are less than provided
+                threshold.
+
+        Returns:
+            recognized_signers: set of "name" of "keys" for which verification succeeds.
+        """
+
+        recognized_signers = []
+        used_keyids = []
+        pae = self.pae
+
+        # checks for threshold value
+        if len(keys) < threshold or threshold <= 0:
+            raise ValueError(
+                f"not a valid threshold, must be between 0 and {len(keys)}"
+            )
+
+        for signature in self.signatures:
+            # Loop over the keys.
+            for (name, key) in keys:
+                if not isinstance(key, Key):
+                    raise TypeError(f"expected type 'Key', got {type(key).__name__}")
+
+                # If key and signature include keyIDs but do not match skip.
+                if (
+                    signature.keyid is not None
+                    and key.keyid is not None
+                    and signature.keyid != key.keyid
+                ):
+                    continue
```

Comment:
@lukpueh do we want to change the behavior of `ANYKEY_SCHEMA` and the behavior of [verify_signature](https://github.com/in-toto/securesystemslib/blob/d3f48889595c6252ad30ccbec2b4c7d87795e37d/securesystemslib/keys.py#L725)?
I am not so familiar with this code and I can't comment.

Reply:
I don't think it will cause circular imports. I have two methods in mind, one is that you suggested.
1. To implement verification in `SSlibKey`, I suggest not to copy the whole `verify_signature()` instead split it into half and reuse that part, maybe here:
https://github.com/in-toto/securesystemslib/blob/d3f48889595c6252ad30ccbec2b4c7d87795e37d/securesystemslib/keys.py#L826-L830
because after this point there is no use of `keyid` in verification. Similarly, for `create_signature()` method this have to be done.
2. The other solution is by making `keyid = Schema.Optional(KEYID_SCHEMA)` in all schema such as `ANYKEY_SCHEMA` and `KEYID_SCHEMA` have to support empty string since it's optional.
https://github.com/in-toto/securesystemslib/blob/d3f48889595c6252ad30ccbec2b4c7d87795e37d/securesystemslib/formats.py#L221-L228

OR, we can leave this for now because in-toto/tuf metadata doesn't have keyid as optional currently.

Reply:
I think I'm in favour of opening an issue for it and moving it out of this PR. Can you open an issue?

Reply:
Submitted an issue #416.